### DISCUSSION TITLE
Just another question about best practice for template inheritance.

### DISCUSSION THREAD

**Discussion by huxuan** [copier-org/copier#1274] [Q&A] [answered]
Dear Copier Community,

Thanks for creating the excellent tool for rendering templates. I notice that there are several discussions about inheritance [1][2] and several potential solutions or references [3][4]. But seriously, those may not meet my needs.

I want to create a series of templates, all based on Python. So the base template is a lightweight Python template, creating an empty Python project with some common components, e.g., lint, CI/CD. And then I want to extend it for multiple specific scenarios. For example, A FastAPI template for developing backend services, A PyTorch template for developing deep learning algorithms. There are several features I want:

1. The extended template itself is a complete template. So multiple templates or including other yaml files may not be OK as in that way, the extended template tends to only contain components not included in the base template.
2. The extended template can be updated alongside base template. If there are some update in base template, it would be better if I do not need to duplicate the changes in all the extended templates.

Currently, the only approach I can come up with is using a GitHub Fork [5] and rebase/merge base template from time to time. But I can imagine there will be more and more conflicts in the future. Just want to know if there is any other best practice. Let me know if there is any misunderstanding I have.

Thanks again for the excellent tools.

[1] https://github.com/orgs/copier-org/discussions/402
[2] https://github.com/orgs/copier-org/discussions/416
[3] https://copier.readthedocs.io/en/stable/configuring/#include-other-yaml-files
[4] https://copier.readthedocs.io/en/stable/configuring/#applying-multiple-templates-to-the-same-subproject
[5] https://docs.github.com/en/get-started/quickstart/fork-a-repo

**Comment by huxuan** [comment]
Some further explanation for try to have the extended template to be complete.

1. Avoid duplicate answers for same questions.
2. Make it easy to test and verify the extended template.
3. Make it easy (for ordinary template users) to update the generated projects.

**Comment by yajo** [comment, marked as answer]
Maybe you need a meta-template? Check out https://github.com/copier-org/copier/issues/934#issuecomment-1405492635.

**Reply by huxuan** [reply]
Wow, that is pretty cool! Thanks @yajo!

**Reply by huxuan** [reply]
Hi @yajo, just want to double confirm about your preference after looking through the long discussion. So meta template is not what you preferred actually, right? Maybe I should keep using git rebase until it becoming a disaster?

**Reply by yajo** [reply]
They are different strategies with different pros and cons. For the use case in that discussion, the Git fork ended up being best. But meta-templates are totally fine too. I just have to find the time to document all these strategies appropriately. 😅

> Maybe I should keep using git rebase until it becoming a disaster?

If you follow this path, use merge instead of rebase, so you keep replayability.
